core/src/main/java/org/apache/calcite/rel/rules/ProjectJoinTransposeRule.java

```diff
@@ -142,13 +144,9 @@ public void onMatch(RelOptRuleCall call) {
 
     // create a new join with the projected children
     Join newJoinRel =
-        join.copy(
```

Comment:
Creating new operators should always pass by the specialized `RelBuilderFactory` which should be responsible for doing things right.

Reply:
Well, i agree, use `RelBuilder` to construct the relational expression.

Reply:
There's a reason we use `RelNodeSubClass.copy`. We need to match the calling convention of the RelNode being copied. 

I couldn't find a JIRA case for removing AbstractRelNode.copy, but this is closely related: https://issues.apache.org/jira/browse/CALCITE-2064.

Reply:
I remember we had a discussion before about this. Because for rules that can match both logical and physical operators, `copy` method can keep it. `RelBuilderFacory` will most likely create logical relational nodes. I don't mind the default rule instance matching logical rel nodes only.

Reply:
For both way, i have no strong objections, since we have change this rule to match only logical nodes, i think `#copy` is not that necessary.

Reply:
It seems that the reason for #EnumerableMergeJoin.copy throwing AssertionError is that the #join.copy use #join.getTraitSet() directly. But the collation may have offset since the new join has new left and new right. So the RelCollation of the TraitSet should be update. I am working for that.

Reply:
I updated the collations when creating a copy of the join with the new inputs.

Reply:
> Can you explain more, although the new join has new inputs, their type doesn't change, right ? So what do you mean by the offset ?

For example: 
EnumerableProject(age=[$3])
  EnumerableMergeJoin(condition=[=($1, $2)], joinType=[inner])
    EnumerableValues(tuples=[[{ '1', 'anna' }, { '2', 'bob' }, { '3', 'tom' }]])
    EnumerableValues(tuples=[[{ 'anna', '14' }, { 'bob', '17' }, { 'tom', '22' }]])
before Project be pushed down, the collations of MergeJoin is [1, 2] and the leftkey of the JoinInfo is [1], 
after Project be pushed down, the left child of MergeJoin becomes LogicalProject#68 whose output rowtype is RecordType(CHAR(4) name), so the leftkey becomes [0] (The index of field was changed). So IMO, the collations of new Join should be updated due to changes of the inputs.

Reply:
Thanks for the explain, @wenhuitang , then fixing the merge join is more acceptable.

Reply:
Thanks @zabetak , it seems that only SMJ and the probe of HashJoin can keep the collation of their inputs, so can we fix the #copy of the EnumerableMergeJoin and EnumerableHashJoin instead of modify the rule ?

Reply:
@danny0405  I don't think we should change the `copy` methods. The contract of copy says that we should be able to modify traits and inputs (check my [previous comment](https://github.com/apache/calcite/pull/1462#discussion_r327851499)).

Reply:
Okey, it seems to late to make the fix into the copy, because all the inputs collations are already wrong, but i still have strong objections on fixing physical traits in logical plans.